kanidm_tools/src/cli/account.rs

```diff
@@ -189,37 +190,81 @@ impl AccountOpt {
+                        if let Err(e) = input_result {
+                            eprintln!("Failed to read from stdin -> {:?}", e);
+                            break;
+                        };
 
-                    // Convert to a u32.
-                    let totp = match u32::from_str_radix(totp_input.trim(), 10) {
-                        Ok(v) => v,
-                        Err(e) => {
-                            eprintln!("Invalid TOTP -> {:?}", e);
-                            return;
-                        }
-                    };
+                        // Convert to a u32.
+                        let totp = match u32::from_str_radix(totp_input.trim(), 10) {
+                            Ok(v) => v,
+                            Err(e) => {
+                                eprintln!("Invalid TOTP -> {:?}", e);
+                                // Try again.
+                                continue;
+                            }
+                        };
 
-                    match client.idm_account_primary_credential_verify_totp(
-                        acsopt.aopts.account_id.as_str(),
-                        totp,
-                        session,
-                    ) {
-                        Ok(_) => {
-                            println!("TOTP registration success.");
-                        }
-                        Err(e) => {
-                            eprintln!("Error Completing -> {:?}", e);
+                        match client.idm_account_primary_credential_verify_totp(
+                            acsopt.aopts.account_id.as_str(),
+                            totp,
+                            session,
+                        ) {
+                            Ok(_) => {
+                                println!("TOTP registration success.");
+                                break;
+                            }
+                            Err(ClientError::TotpInvalidSha1(session)) => {
+                                eprintln!("⚠️  WARNING - It appears your authenticator app may be broken ⚠️  ");
+                                eprintln!(" The TOTP authenticator you are using is forcing the use of SHA1");
+                                eprintln!("");
+                                eprintln!(" -- If you accept this risk, and wish to proceed, type 'I am sure' ");
+                                eprintln!(" -- Otherwise press ENTER to cancel this operation");
+                                eprintln!("");
+                                eprint!("Are you sure: ");
+
+                                let mut confirm_input = String::new();
+                                if let Err(e) = io::stdin().read_line(&mut confirm_input) {
+                                    eprintln!("Failed to read from stdin -> {:?}", e);
+                                    break;
+                                };
+
+                                if confirm_input.trim() == "I am sure" {
```

Comment:
Should that be case sensitive?

Reply:
I'd argue yes - you really want them to type it properly, or copy-paste it like every one of these things I come across 😆

Reply:
I think even just at a point they type 'i am sure' that's still strong enough "intent" that insensitive here is probably fine. We just need it to be a deliberate and intended action, rather than an accidental accept.